### DISCUSSION TITLE
unwrapping reference point

### DISCUSSION THREAD

**Discussion by rita966** [isce-framework/isce2#464] [Q&A] [answered]
i used topsApp.py to do dinsar
and how can i know the unwrapping reference point of snaphu_mcf ?
or can i set the reference point by myself?
thanks!

**Comment by gmgunter** [comment, marked as answer]
Hi @eric890610. 

Just for clarity, I assume that the "unwrapping reference point" refers to a pixel in the unwrapped phase data that should match the wrapped phase of the input interferogram (without any additional cycles of 2pi). SNAPHU does not expose this as a configurable parameter, AFAIK.

SNAPHU estimates the unwrapped phase field indirectly by first solving for the unwrapped phase "gradients" (the first-order discrete differences between neighboring phase values) and then integrating them along a path that spans the interferogram, starting from some seed pixel. This initial seed pixel is the only pixel in the unwrapped result that will reliably match the corresponding wrapped phase value.

Looking at the source code, SNAPHU seems to use the [first (i.e. upper-left corner) pixel in the output grid as the seed pixel](https://github.com/isce-framework/isce2/blob/566ebfb9a85f2bdb5d538b3ba86b289aaaf702a7/contrib/Snaphu/src/snaphu_util.c#L278-L279), but this seems like sort of an implementation detail of the software -- not necessarily an explicit guarantee that you should rely on in your own code.

If you want, it seems straightforward to adjust the unwrapped phase values as a post-processing step to use some other pixel (i, j) as a new reference pixel. You would simply compute the following value and subtract it from each pixel in the unwrapped phase data:

<p align="center">
<img src="https://latex.codecogs.com/svg.image?2\pi&space;\left\lceil&space;\frac{\Phi(i,&space;j)&space;-&space;\pi}{2\pi}&space;\right\rceil">
</p>

where <img src="https://latex.codecogs.com/svg.image?\Phi(i,&space;j)"> is the unwrapped phase value at the desired reference point, and <img src="https://latex.codecogs.com/svg.image?\lceil\cdot\rceil"> is the [ceil function](https://en.wikipedia.org/wiki/Floor_and_ceiling_functions). This operation will shift the unwrapped phase values such that the value of pixel (i, j) is within the interval (-pi, pi], while preserving the relative phase between pixels.

**Comment by hfattahi** [comment]
Thanks @gmgunter for the clear description. 
@eric890610 it is very common in the post processing we choose a reference pixel and subtract the unwrapped value of that pixel from the unwrapped phase of all pixels. In practice if you know your area, you may choose a reference pixel in a region with high correlation and non-zero connected component. Many of these post-processing methods have been automated in time-series software such as MintPY.
